src/catalog/src/durable/error.rs

```diff
@@ -119,3 +130,102 @@ impl From<TryFromProtoError> for DurableCatalogError {
         DurableCatalogError::Proto(e)
     }
 }
+
+/// An error that indicates the durable catalog has been fenced.
+///
+/// The order of this enum indicates the most information to the least information.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, thiserror::Error)]
+pub enum FenceError {
+    /// This instance was fenced by another instance with a higher deployment generation. This
+    /// necessary means that the other instance also had a higher epoch. The instance that fenced
+    /// us believes that they are from a later generation than us.
+    #[error("current catalog deployment generation {current_generation} fenced by new catalog epoch {fence_generation}")]
+    DeployGeneration {
+        current_generation: u64,
+        fence_generation: u64,
+    },
+    /// This instance was fenced by another instance with a higher epoch. The instance that fenced
+    /// us may or may not be from a later generation than us, we were unable to determine.
+    #[error("current catalog epoch {current_epoch} fenced by new catalog epoch {fence_epoch}")]
+    Epoch {
+        current_epoch: Epoch,
+        fence_epoch: Epoch,
+    },
+    // It's very likely that if we were to read the latest updates after this error, then we'd see
+    // a higher epoch/deploy generation from the other writer and get fenced through that. We could
+    // potentially always do that and eliminate this fence variant. However, the catalog debug tool
+    // is able to write to the catalog without incrementing the epoch and therefore would not be
+    // able to fence this instance. It's technically possible for this instance to survive and
+    // respond to a write from the catalog debug tool, but until more thought is put into it, the
+    // safest thing to do is crash and reboot.
```

Comment:
I've realized that this isn't actually true. It's possible for the following to happen:

1. Catalog opens.
2. Catalog debug tool writes something.
3. Catalog performs a read, which first involves listening to all changes in the shard, including the changes written by the debug tool. This also updates the internal upper. 
4. Catalog goes to make a write, it's already updated it's internal upper so no upper mismatch happens.

Reply:
This is an existing issue though so I'll address it in a follow up.